### PR TITLE
[network] prefer interface with gateway

### LINKED ISSUE #14964: Wake on LAN only broadcasts on one network interface
### Describe the bug
I wanted to make my PVR backend go to sleep when idle. MythTV provides this functionality, and the PVR add-on for Kodi supports Wake-on-LAN to facilitate this. This functionality appears to be part of Kodi (the addon calls XBMC->WakeOnLan with the MAC address) so I believe this to be related to Kodi rather than the specific add-on.

First, I tested by manually forcing the backend host to go to sleep (actually suspend to RAM), then using a Wake-on-LAN tool on my router to power on the host via its MAC address. That worked.

Then I tested Kodi on Windows: I input the MAC address into the PVR add-on settings, forced the backend to sleep, restarted Kodi and accessed the PVR functionality - the backend host remained asleep. I waited a few minutes and eventually decided to wake it up via my router's WOL tool again (which worked.)

Wireshark (on the Windows PC) showed that the WOL packet was being sent on only one of the network interfaces (I have VirtualBox on this PC and it was showing on that interface). No WOL packet was being sent on my physical network interface according to Wireshark.

After disabling the VirtualBox ethernet interface, Kodi was able to wake up the remote host, indicating that Kodi does send the packet and that the receiving system is able to respond correctly to the packets (by waking up), however the packets might be sent on the wrong interface.

It seems that the code in xbmc/network/Network.cpp sends the packet to the INADDR_BROADCAST address, and reading about this online it seems this will send a single packet on only one (unspecified) network interface.

## Expected Behavior
Kodi's Wake-on-LAN should wake up the host whose network card has the specified MAC address.

## Actual Behavior
The remote host remains asleep if the networking stack chooses a different network interface to broadcast the WOL packet on.

## Possible Fix
Send one WOL packet per available network interface: Loop through each network interface that has been assigned an IP address, identify the broadcast address for each of these and send the WOL packet to those instead of the single packet to the INADDR_BROADCAST address.

### To Reproduce
Ensure your computer has multiple network interfaces present (either physical, or some virtual ones e.g. VM interfaces or VPN links) and enabled.
Try using Kodi to wake up a machine on the same network via Wake-on-LAN (e.g. install MythTV PVR add-on and put the MAC address of that computer in the settings and try to access the PVR functionality).
Note: This should be reproducible with any Kodi component that uses Wake-on-LAN, I have only had experience with this functionality via the specific add-on I refer to though.

### Debuglog (added 28th Nov 2018)
Looking at the code it doesn't look like it will give any useful information in this case (only that it successfully sent the WOL packet - not which interface it sent it on.)

Debug log from Windows: https://paste.kodi.tv/ogawupadit.kodi

Relevant lines:
21:29:26.704 T:2948    INFO: CNetworkBase::WakeOnLan - Magic packet send to '00:25:90:98:38:0e'
21:29:29.715 T:2948    INFO: CNetworkBase::WakeOnLan - Magic packet send to '00:25:90:98:38:0e'

Wireshark packet example (for first WOL packet) from 192.168.56.1 ("VirtualBox Host-Only Network" interface: https://paste.kodi.tv/avapimohuv

### Screenshots 
n/a

## Additional context or screenshots (if appropriate)
n/a

### Your Environment
Used Operating system:
 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberri-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10
 - Kodi version: 18 / Leia

### PR DESCRIPTION
**Pull request by Rechi**
## Motivation and Context
show the correct network information in system info if you have multiple NICs (e.g. VirtualBox)

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/network/Network.cpp [hunk 1]
```diff
@@ -202,17 +202,19 @@ bool CNetworkBase::IsLocalHost(const std::string& hostname)
 
 CNetworkInterface* CNetworkBase::GetFirstConnectedInterface()
 {
-   std::vector<CNetworkInterface*>& ifaces = GetInterfaceList();
```
Comment (FernetMenta) [inline xbmc/network/Network.cpp]: how does this fix the issue? I don't see this method called by CNetworkBase::WakeOnLan (maybe it is there and I am just blind)

Comment (Rechi) [reply xbmc/network/Network.cpp]: Yes you are right it doesn't fix it. I thought `WakeOnLan` is implemented in NetworkInterface but it is implemented in Network and mac address lookup iterates over all interfaces.
Next time I actually should look at the code.

**Comment by wsnipex** [comment]
is this still waiting on something?

**Comment by wsnipex** [comment]
yes, looks sensible. Maybe remove the fixes: 14964 and merge

**Pull request merged**